sw/device/lib/dif/dif_pwrmgr.c

```diff
@@ -4,8 +4,543 @@
+}
+
+/**
+ * Checks if a value is a valid `dif_pwrmgr_req_type_t`.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool is_valid_req_type(dif_pwrmgr_req_type_t val) {
+  return val == kDifPwrmgrReqTypeWakeup || val == kDifPwrmgrReqTypeReset;
+}
+
+/**
+ * Checks if a value is a valid `dif_pwrmgr_irq_t`.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool is_valid_irq(dif_pwrmgr_irq_t val) {
+  return val >= kDifPwrmgrIrqFirst && val <= kDifPwrmgrIrqLast;
+}
+
+/**
+ * Checks if a value is valid for, i.e. fits in the mask of, a
+ * `bitfield_field32_t`.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool is_valid_for_bitfield(uint32_t val, bitfield_field32_t bitfield) {
+  return (val & bitfield.mask) == val;
+}
+
+/**
+ * Checks if the control register is locked.
+ *
+ * Control register is locked when low power is enabled and WFI occurs. It is
+ * unclocked when the chip transitions back to active power state.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool control_register_is_locked(const dif_pwrmgr_t *pwrmgr) {
+  if (!bitfield_bit32_read(
+          mmio_region_read32(pwrmgr->params.base_addr,
+                             PWRMGR_CTRL_CFG_REGWEN_REG_OFFSET),
+          PWRMGR_CTRL_CFG_REGWEN_EN)) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * The configuration registers CONTROL, WAKEUP_EN, and RESET_EN are all written
+ * in the fast clock domain but used in the slow clock domain. Values of these
+ * registers are not propogated across the clock boundary until this function is
+ * called.
+ */
+static void sync_slow_clock_domain(const dif_pwrmgr_t *pwrmgr) {
+  // Start sync and wait for it to finish.
+  mmio_region_write32(pwrmgr->params.base_addr, PWRMGR_CFG_CDC_SYNC_REG_OFFSET,
+                      bitfield_bit32_write(0, PWRMGR_CFG_CDC_SYNC_SYNC, true));
+  while (bitfield_bit32_read(mmio_region_read32(pwrmgr->params.base_addr,
+                                                PWRMGR_CFG_CDC_SYNC_REG_OFFSET),
+                             PWRMGR_CFG_CDC_SYNC_SYNC)) {
+  }
+}
+
+/**
+ * Checks if sources of a request type are locked.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool request_sources_is_locked(const dif_pwrmgr_t *pwrmgr,
+                                      dif_pwrmgr_req_type_t req_type) {
+  request_reg_info_t reg_info = request_reg_infos[req_type];
+  uint32_t reg_val = mmio_region_read32(pwrmgr->params.base_addr,
+                                        reg_info.write_enable_reg_offset);
+  // Locked if write enable bit is set to 0.
+  return !bitfield_bit32_read(reg_val, reg_info.write_enable_bit_index);
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_init(dif_pwrmgr_params_t params,
+                                    dif_pwrmgr_t *pwrmgr) {
+  if (pwrmgr == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  *pwrmgr = (dif_pwrmgr_t){.params = params};
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_config_result_t dif_pwrmgr_low_power_set_enabled(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t new_state) {
+  if (pwrmgr == NULL || !is_valid_toggle(new_state)) {
```

Comment:
What I've been doing to convert toggle_t into a bool is something like this:

```
bool flag;
switch (state) {
  case kEnabled: flag = true; break;
  case kDisabled: flag = false; break;
  default: return kBad;
}
```
I'm not saying you have to do this, but I feel it helps make the check and conversion clearer, since it all happens in one place.

Reply:
Thought about the same while reviewing the recent i2c DIF changes.